Various underline (and strikethrough) improvements

qwerasd205 opened the pull request:
Improvements to underline and strikethrough rendering, including:
- Reworked the underline sprite font rendering for saner results in terms of positioning
- Adjusted strikethrough position metric calculations
- Made underlines draw *before* text, so that they don't harm legibility as much for text with descenders

## Comparisons:

### CoreText
|main|this PR|
|-|-|
|<img width="257" alt="image" src="https://github.com/user-attachments/assets/97b0c975-8019-4fe0-83fa-875163be5743">|<img width="263" alt="image" src="https://github.com/user-attachments/assets/9307b4ee-fab6-4084-b8dd-b6fca67afb4e">|

### FreeType
|main|this PR|
|-|-|
|<img width="264" alt="image" src="https://github.com/user-attachments/assets/ff752254-af54-409b-aadf-f4354a146a31">|<img width="267" alt="image" src="https://github.com/user-attachments/assets/37d841cf-05ea-4f21-9e91-fbdd0d9877fb">|

### TextEdit
<img width="238" alt="image" src="https://github.com/user-attachments/assets/f1c64075-3dbf-4209-97cd-b99e74f6eed3">

qwerasd205 committed:
font/sprite: fix 1px gap at right edge of dotted and dashed underlines

qwerasd205 committed:
font/sprite: rework underline rendering, adjust positioning

qwerasd205 committed:
renderer: draw underlines below text to improve legibility

qwerasd205 committed:
renderer/OpenGL: use better logic for whether to render glyph

Metal already had this change made, so I copied it over from there. This
logic is more straightforward. Also copied the check to skip 0-sized
glyphs, since sometimes, for example, spaces are emitted as glyphs by
the shaper for some reason, even though they have no actual content, and
we want to avoid sending a bunch of useless stuff to the GPU.

qwerasd205 committed:
coretext: improve strikethrough position calculation

qwerasd205 committed:
freetype: improve strikethrough position guess by using ex height

qwerasd205 committed:
font/sprite: reduce uneven gaps in dotted underline

qwerasd205 committed:
font: further improve ul/st position calculations

qwerasd205 committed:
freetype: fix underline position calculation

mitchellh reviewed (changes requested):
This all looks very good, I have a few very minor pieces of feedback.

Besides this, are you happy with this? This seems like a good improvement thank you.

qwerasd205 committed:
correct diagram/description (+Y is down not up)

qwerasd205 committed:
font/sprite: correct underline placement calculation

qwerasd205 committed:
font: remove fudge factors in ul and st position calculations

These were present because of an incorrect calculation in the underline
sprite renderer, and are no longer necessary.

qwerasd205 committed:
freetype: update expected ul pos in tests to account for removed fudge factor
